trio/_subprocess/linux_waitpid.py

```diff
@@ -0,0 +1,85 @@
+import attr
+import functools
+import os
+import outcome
+from typing import Dict, Any
+
+from .. import _core
+from .._sync import Event
+from .._threads import run_sync_in_worker_thread
+
+# type: Dict[int, WaitpidResult]
+_pending_waitpids = {}
+
+
+@attr.s
+class WaitpidResult:
+    event = attr.ib(default=attr.Factory(Event))
+    outcome = attr.ib(default=None)
+    _cached_result = attr.ib(default=None)
+
+    def unwrap(self):
+        if self._cached_result is not None:
+            if isinstance(self._cached_result, BaseException):
+                raise self._cached_result
+            return self._cached_result
+
+        try:
+            self._cached_result = self.outcome.unwrap()
+            return self._cached_result
+        except BaseException as e:
+            self._cached_result = e
+            raise e from None
+
+
+# https://github.com/python-trio/trio/issues/618
+class StubLimiter:
+    def release_on_behalf_of(self, x):
+        pass
+
+    async def acquire_on_behalf_of(self, x):
+        pass
+
+
+waitpid_limiter = StubLimiter()
+
+
+# adapted from
+# https://github.com/python-trio/trio/issues/4#issuecomment-398967572
+async def _task(pid: int) -> None:
+    """The waitpid thread runner task. This must be spawned as a system
+    task."""
+    partial = functools.partial(
+        os.waitpid,  # function
+        pid,  # pid
+        0  # no options
+    )
+    try:
+        tresult = await run_sync_in_worker_thread(
+            outcome.capture,
+            partial,
+            cancellable=True,
+            limiter=waitpid_limiter
+        )
+    except Exception as e:
+        result = _pending_waitpids.pop(pid)
+        result.outcome = outcome.Error(e)
+        result.event.set()
+        raise
```

Comment:
The only way to hit this `except` block (modulo bugs) is if the `run_sync_in_worker_thread` gets cancelled. Cancellation is always a `BaseException`, not an `Exception`, so right now I think this is a no-op? Also, 

It's really not clear whether we even care about cancellation here... since this is a system task, the only way it can get cancelled is if the entire `run` is shutting down, either because the main task has already finished, or because there was an internal error and we're crashing. So... I guess the only time it matters whether we call `result.event.set()` here is when someone is doing a *cancel-shielded* wait for a subprocess to finish? And if they do, we won't handle it correctly *anyway* – the user would expect a cancel-shielded wait to actually wait for the subprocess to exit, but we'll still abort the `run_sync_in_worker_task`, so we won't know when the subprocess exits. Maybe we need to resurrect #303 to handle this corner case? Or toggle the system task's shielding on and off depending on whether anyone is currently blocked in `Popen.wait`? Or use `threading.Thread` directly instead of trying to go through `run_sync_in_worker_task`? (...can we even reliably re-enter the trio thread once the system nursery is cancelled? The re-entry queue processor is also a system task...)

Man, I hate `waitpid`.

This is such an obscure use case that I don't think we need to worry about it right now. Trying to do a cancel-shielded wait for a subprocess isn't *too* ridiculous – `nursery.__aexit__` is one of the standard use cases for cancel-shielding, and people will want to do nursery-like things for managing subprocesses. So we might end up caring eventually. But right now we should just shrug and accept that it won't do exactly the right thing here. And if we don't care about that, we can make this whole function a lot simpler, like...

```python
try:
    result.outcome = await run_sync_in_worker_thread(...)
finally:
    # Paranoia to make sure we unblock any waiters even if something goes horribly wrong
    result.event.set()
```

Reply:
> Trying to do a cancel-shielded wait for a subprocess isn't too ridiculous – `nursery.__aexit__` is one of the standard use cases for cancel-shielding, and people will want to do nursery-like things for managing subprocesses.

Actually, on further thought, this is still overstating the importance of this edge case. If someone does a cancel-shielded wait for a subprocess, that will mostly work fine, even with my "simplified" code above. The only way this task gets cancelled is if we're crashing, or if the main task has already exited. That means that the only ways a cancel-shielded `Popen.wait` can fail to actually wait are:

* We're in the middle of crashing with a `TrioInternalError`: well, sorry, that means internal invariants have been violated and we're blowing up the world, so it's OK if your task manager doesn't wait for child processes correctly. At this point all guarantees are void.

* If we're not crashing with `TrioInternalError`, and the system task is cancelled, and someone is doing a cancel-shielded `Popen.wait`, then the main task has already exited, so they must be doing it from *inside a system task*. But doing *anything* inside a cancel-shield in a system task is highly dubious, because by the time a system task gets cancelled the world is being torn down around you. Don't do that please.